### PR TITLE
Fix ItemsSource and TreeViewNode syncing

### LINKED ISSUE #1182: TreeView with ItemTemplate "removes" Items unexpectedly
**Describe the bug**

When using a TreeView with drag/drop enabled and custom ItemPlate with DataBinding, there is a situation where moving a TreeView Item into another may result in losing TreeView item childs when expanding/closing the root TreeView node.

Example repo: https://github.com/chingucoding/TreeViewBugIsolated

**Steps to reproduce the bug**

Steps to reproduce the behavior (use [this repository]():
1.  Start App
2.  Drag Folder "Personal Folder" into folder "Work Documents"
3.  Expand "Personal Folder" so you can see that "Home Remodel Folder" has a expand chivron
4.  Contract the "Work Documents" item.
5.  Expand "Work Documents"
5.  Contract "Work Documents"
6.  Expand "Personal Folder" -> The item "Home Remodel Folder" has no expand chivron.

**Expected behavior**

The item should have an expand chivron.
**Screenshots**

treeviewbug
**Version Info**

NuGet package version: 

2.1.190606001

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Insider Build (18956)              | Yes   |

| Device form factor | Saw the problem? |
| :-------------------- | :------------------- |
| Desktop                 | Yes |

**Additional context**

See [this issue](https://github.com/microsoft/Xaml-Controls-Gallery/issues/94) for additional information.
This bug/issue is present in the current version of the [XAML Controls Gallery app](https://www.microsoft.com/en-us/p/xaml-controls-gallery/9msvh128x2zt)

### PR DESCRIPTION
**Pull request by kaiguo**
Fixes #1182 

In expand/collapse scenario, TreeViewNode.Children already contains the right content for ItemsSource, which means we don't need to clear and re-populate ItemsSource into children collection again. Re-populating for every expand can potentially create an infinite loop because updating TreeViewNode.Children collection triggers ListView layout updates, which can trigger ItemsSource update in some cases (when there are multiple levels of children and TreeViewItems got recycled), if we clean and re-populate children collection again it becomes an infinite loop.

Added a check to make sure we only update TreeViewNode children collection when the contents are actually different from ItemsSource.

### CODE REVIEW

File: dev/TreeView/InteractionTests/TreeViewTests.cs [hunk 1]
```diff
@@ -2688,6 +2688,61 @@ public void TreeViewRootNodeBindingTest()
             }
         }
 
+        [TestMethod]
```
Comment (RBrid) [inline dev/TreeView/InteractionTests/TreeViewTests.cs]: Looks like the InteractionTests.TreeViewTests.TreeViewSwappingNodesTest_ContentMode test regressed.

Comment (kaiguo) [reply dev/TreeView/InteractionTests/TreeViewTests.cs]: > Looks like the InteractionTests.TreeViewTests.TreeViewSwappingNodesTest_ContentMode test regressed.

Just saw your reviews...
The test is setting ItemsSource to null and I was accessing itemsSource.Count without any check, which caused crash, should be fixed now.

File: dev/TreeView/TreeViewNode.cpp [hunk 2]
```diff
@@ -248,6 +248,31 @@ void TreeViewNode::SyncChildrenNodesWithItemsSource()
...
+    {
+        return false;
+    }
+
+    // Compare the actual content in collections when counts are equal
+    if (itemsSourceCount > 0)
+    {
+        for (UINT32 i = 0; i < itemsSourceCount; i++)
+        {
+            if (Children().GetAt(i).Content() != m_itemsDataSource.GetAt(i))
```
[the remark this answers is not in the document]
Comment (kaiguo) [reply dev/TreeView/TreeViewNode.cpp]: Updated :)

[the remark this answers is not in the document]
Comment (kaiguo) [reply dev/TreeView/TreeViewNode.cpp]: There's a `if (childrenCount != itemsSourceCount)` check at the top, so we shouldn't get here if GetAt(i) is null.

**Pull request merged**